### PR TITLE
Tdl 681 add retry mechanism

### PR DESCRIPTION
**Pull request by harshpatel4crest**
# Description of change
- TDL-681: HTTP requests that fail are not retrying - Added retry mechanism for requests that are failing
- Added exception handling ie. KlaviyoError

# Manual QA steps
- Manually raised KlaviyoError and JSONDecoder error and checked retrying of requests
- Test Cases that checks the retrying of requests on error
- Test Cases for checking exception handling and error message
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_klaviyo/utils.py [hunk 1]
```diff
@@ -3,13 +3,67 @@
...
+    },
+    500: {
+        "raise_exception": KlaviyoInternalServiceError,
+        "message": "Something is wrong on Klaviyo's end."
+    }
+}
+
+def raise_for_error(response):   
+    try:
+        json_resp = response.json()
```
Comment (karanpanchal-crest) [inline tap_klaviyo/utils.py:58]: @harshpatel4crest Can you please get the response of response.raise_for_status() here and except the HTTTError in the exception.
Please refer to the tap-Xero code (https://github.com/singer-io/tap-xero/pull/85/files#diff-43d32c99c6992e01f4dbdb49df5e372db029dae0de306b9191283ef21b852438R216) for the same

File: tests/test_exception_handling.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
+import tap_klaviyo.utils as utils_
+import unittest
+from unittest import mock
+import requests
+import singer
+
+class Mockresponse:
+    def __init__(self, resp, status_code, content=None, headers=None, raise_error=False):
```
Comment (karanpanchal-crest) [inline tests/test_exception_handling.py:8]: @harshpatel4crest You can make resp an optional argument. 
`def __init__(self, status_code, resp={}, content=None, headers=None, raise_error=False)

def klaviyo_400_error(*args, **kwargs):
    return Mockresponse(json_str, 400, raise_error=True)
`

Please follow above code for remaining functions

File: tests/test_exception_handling.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+def klaviyo_500_error(*args, **kwargs):
+    json_str = {}
+
+    return Mockresponse(json_str, 500, raise_error=True)
+
+
+class TestBackoff(unittest.TestCase):
+    
+    @mock.patch('requests.Session.request', side_effect=klaviyo_400_error)
+    def test_400_error(self, klaviyo_400_error):
```
Comment (karanpanchal-crest) [inline tests/test_exception_handling.py:61]: @harshpatel4crest Can you also add a positive scenario as well

File: tap_klaviyo/utils.py [hunk 4]
```diff
@@ -3,13 +3,70 @@
...
+        "raise_exception": KlaviyoBadRequestError,
+        "message": "Request is missing or has a bad parameter."
+    },
+    401: {
+        "raise_exception": KlaviyoUnauthorizedError,
+        "message": "Invalid authorization credentials."
+    },
+    403: {
+        "raise_exception": KlaviyoForbiddenError,
+        "message": "Request is missing or has an invalid API key."
```
Comment (KAllan357) [inline tap_klaviyo/utils.py:44]: Can we adjust the message here to say "Invalid authorization credentials or permissions."

File: tap_klaviyo/utils.py [hunk 5]
```diff
@@ -3,13 +3,70 @@
...
+    },
+    404: {
+        "raise_exception": KlaviyoNotFoundError,
+        "message": "The requested resource doesn't exist."
+    },
+    500: {
+        "raise_exception": KlaviyoInternalServiceError,
+        "message": "Something is wrong on Klaviyo's end."
```
Comment (KAllan357) [inline tap_klaviyo/utils.py:52]: Can we also adjust this to say "Internal Service Error from Klaviyo"

**Review by KAllan357 — CHANGES_REQUESTED** [review]
Please change some of the error message text.

File: tap_klaviyo/utils.py (hunk 4 above)
Comment (harshpatel4crest) [reply tap_klaviyo/utils.py]: I have added the desired string, however I have taken the string from Klaviyo Documentation https://www.klaviyo.com/docs/api/lists.
So, let me know if want to revert the change.

File: tap_klaviyo/utils.py (hunk 5 above)
Comment (harshpatel4crest) [reply tap_klaviyo/utils.py]: Added the string in message

**Review by KAllan357 — APPROVED** [review]

**Comment by KAllan357** [comment]
Can you resolve the conflicts?

**Comment by harshpatel4crest** [comment]
> Can you resolve the conflicts?

@KAllan357 
Resolved conflicts

**Pull request merged**